staging/src/k8s.io/legacy-cloud-providers/azure/auth/azure_auth.go

```diff
@@ -125,13 +132,25 @@ func GetServicePrincipalToken(config *AzureAuthConfig, env *azure.Environment) (
 	return nil, ErrorNoAuth
 }
 
-// ParseAzureEnvironment returns azure environment by name
-func ParseAzureEnvironment(cloudName string) (*azure.Environment, error) {
+// ParseAzureEnvironment returns the azure environment.
+// If 'cloudFQDN' is set, environment is computed by quering the cloud's resource manager endpoint.
+// Otherwise, a pre-defined Environment is looked up by name.
+func ParseAzureEnvironment(cloudName, cloudFQDN, identitySystem string) (*azure.Environment, error) {
 	var env azure.Environment
 	var err error
-	if cloudName == "" {
+	if cloudFQDN != "" {
+		resourceManagerEndpoint := fmt.Sprintf("https://management.%s/", cloudFQDN)
+		nameOverride := azure.OverrideProperty{Key: azure.EnvironmentName, Value: cloudName}
+		klog.V(4).Infof("Loading environment from resource manager endpoint: %s", resourceManagerEndpoint)
+		env, err = azure.EnvironmentFromURL(resourceManagerEndpoint, nameOverride)
+		if err == nil && strings.EqualFold(cloudName, "AzureStackCloud") {
```

Comment:
Do we have to use AzureStackCloud name? Can we use something more generic like CustomCloud or something along those lines?

Reply:
The dynamic endpoints has also become now a requirement for  sovereign, gov clouds and we also want to be able to proxy it out for our e2e test mocking.

Reply:
can remove `strings.EqualFold(cloudName, "AzureStackCloud")`